README.md
# Práctica 2 - Seguridad y criptografía.
Iris Álvarez Nieto y Carmen Díez Menéndez. Pareja 3.

## Decisiones de diseño y funcionamiento

Se incluyen en la [Wiki](../../wikis/home).

## Instrucciones para realizar la prueba

### Datos necesarios
- **NIA**: 379789
- **Token**: 6213f5aCF0Ad8EBb

### Comandos usado y a ejecutar

```console
foo@bar:~$ python3 securebox_client.py --upload prueba2.txt --dest_id e281430
foo@bar:~$ python3 securebox_client.py --download cB9DdF0E --source_id 379789
```




changes.md
## Mejoras
- Ampliadas decisiones de diseño y seguridad sobre el fichero de configuración (apartado de **Configuración inicial y seguridad del token**).
- Arreglado `list_files` (no pide `source_id`).
- Error con `create_id` y los bytes arreglado instalando [simplejson](https://simplejson.readthedocs.io/en/latest/).

## Extras
- Protegemos el fichero de configuración, y con él el token, encriptándolo con una contraseña. Guardamos su hash para comprobarla cada vez que el usuario use la aplicación.
- Añadimos los endpoints de la API al fichero de configuración.


user.py
import requests
from Crypto.PublicKey import RSA
import utils as u

u.config_ini()
headers = {'Authorization': "Bearer " + u.token}
urlIni = 'https://tfg.eps.uam.es:8080/api/users/'

def create_id(nombre, email):
    global urlIni, headers
    # Creamos el par de claves
    print("Generando par de claves RSA de 2048 bits")
    key = RSA.generate(2048)
    privateKey = key.export_key()
    publicKey = key.publickey().export_key()
    url = urlIni + 'register'
    args = {'nombre': nombre, 'email': email, 'publicKey': publicKey}
    
    try:
        r = requests.post(url, headers = headers, json = args)
    
    except requests.ConnectionError:
        print("Error de conexion")
        return
    
    if r.status_code == requests.codes.ok:
        # Guardamos la clave privada ya que el usuario se crea correctamente
        #TODO ver como guardar 
        #file_out = open("private.pem", "wb")
        #file_out.write(private_key)
        answer = r.json()
        userID = answer.get('userID')
        print("Identidad con ID# {} creada correctamente".format(userID))
    else:
        u.error(r)
    

def search_id(data_search):
    global urlIni, headers
    url = urlIni + 'search'
    args = {'data_search': data_search}
    print("Buscando usuario '" + data_search + "' en el servidor")
    try:
        r = requests.post(url, headers = headers, json = args)
    except requests.ConnectionError:
        print("Error de conexion")
        return
    
    if r.status_code == requests.codes.ok:
        answers = r.json()
        print("{} usuarios encontrados:".format(len(answers)))
        i = 0
        for answer in answers:
            print("[{}] {}, {}, ID: {}".format(i+1, answers[i]['nombre'], answers[i]['email'], answers[i]['userID']))
            i += 1
    else:
        u.error(r)

def delete_id(userID):
    global urlIni, headers
    url = urlIni + 'delete'
    args = {'userID': userID}
    print("Solicitando borrado de la identidad #{}".format(userID))

    try:
        r = requests.post(url, headers = headers, json = args)
    except requests.ConnectionError:
        print("Error de conexion")
        return
    
    if r.status_code == requests.codes.ok:
        answer = r.json()
        userID = answer.get('userID') #TODO poco logico maybe porque argumento pero meeh
        print("Identidad con ID#{} borrada correctamente".format(userID))
    else:
        u.error(r)

def get_public_key(userID):
    global urlIni, headers
    url = urlIni + 'getPublicKey'
    args = {'userID': userID}
    print("Solicitando clave de la identidad #{}".format(userID))

    try:
        r = requests.post(url, headers = headers, json = args)
    except requests.ConnectionError:
        print("Error de conexion")
        return
    if r.status_code == requests.codes.ok:
        print("Clave publica conseguida")
        return r.json().get('publicKey')
    else:
        u.error(r)
        

def prueba():
    create_id("Carmen", "carmen.diezmenendez@estudiante.uam.es")
    search_id("Carmen")
    #delete_id(383336)
    #search_id("Carmen")
    delete_id(38333336)

prueba()


files.py
import requests
import utils as u
import os
import users
import crypto


def upload(file_path, dest_id):
    """
        Nombre: upload
        Descripcion: Envia un fichero a otro usuario.
            Subimos el archivo a SecureBox firmado y cifrado con las claves adecuadas
            para que pueda ser recuperado y verificado por el destinatario.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -dest_id: ID del user que recibira el archivo.
            -file_path: path del archivo a subir.
        Retorno: Ninguno
    """

    print("Solicitando envio de fichero a SecureBox")

    # Abrimos el archivo y vemos donde guardar el cifrado y firmado temporalmente
    try:
        f = open(file_path, "rb")
    except FileNotFoundError:
        print("El archivo no existe")
        return

    mensaje = f.read()
    f.close()
    file_name = os.path.basename(file_path)
    path_archivo = os.path.abspath(u.path_archivos + file_name)

    # Conseguimos la clave publica de dest_id
    clave_pub_r = users.get_public_key(dest_id)
    if clave_pub_r is None:
        print("Error al obtener la clave publica")
        return

    # Firmamos y ciframos el fichero
    mensaje_enc_sign = crypto.enc_sign(mensaje, clave_pub_r)
    if mensaje_enc_sign is None:
        return

    # Guardamos el fichero encriptado y firmado
    f = open(path_archivo, "wb")
    f.write(mensaje_enc_sign)
    f.close()

    # Enviamos la peticion a la API
    print("-> Subiendo fichero a servidor...", end="")
    url = u.url_files + "upload"
    files = {'ufile': open(path_archivo, "rb")}
    try:
        r = requests.post(url, headers=u.headers, files=files)
    except requests.ConnectionError:
        print("\nError de conexion")
        return

    # Si es correcto imprimimos el ID del archivo
    if r.status_code == requests.codes.ok:
        print("OK")
        answers = r.json()
        file_id = answers.get('file_id')
        print("Subida realizada correctamente, ID del fichero: {}".format(file_id))
        # Borramos el archivo auxiliar
        os.remove(path_archivo)
    else:
        print()
        u.error(r)


def download(file_id, source_id):
    """
        Nombre: download
        Descripcion: Recupera un fichero con ID id_fichero del sistema.
            Tras ser descargado, desciframos el contenido y verificamos la firma.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -file_id: ID del fichero en el sistema.
            -source_id: ID del user al que pertenece.
        Retorno: Ninguno
    """
    print("Descargando fichero de SecureBox...", end="")

    # Enviamos la peticion a la API
    url = u.url_files + "download"
    args = {'file_id': file_id}
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("Error de conexion")
        return

    # Si es correcto recuperamos el mensaje
    if r.status_code == requests.codes.ok:
        print("OK")
        print("-> {} bytes descargados correctamente".format(r.headers.get('Content-Length')))

        mensaje_descifrado = crypto.decrypt(r.content)
        if mensaje_descifrado is None:
            return

        # Recuperamos la clave publica del usuario y verificamos la firma
        clave_pub_e = users.get_public_key(source_id)
        if clave_pub_e is None:
            print("Error al obtener la clave publica")
            return

        mensaje_original = crypto.check_sign(mensaje_descifrado, clave_pub_e)
        if mensaje_original is None:
            return

        # Hallamos el nombre del fichero
        aux = r.headers.get('content-disposition')
        nombre = (aux.split("=")[1]).replace('"', '')

        # Guardamos el fichero en path_archivos
        f = open(u.path_archivos + nombre, 'wb')
        f.write(mensaje_original)
        f.close()

        print("Fichero descargado y verificado correctamente")
    else:
        print()
        u.error(r)


def delete(file_id):
    """
        Nombre: delete
        Descripcion: Borra un fichero del sistema.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -file_id: ID del fichero en el sistema.
        Retorno: Ninguno
    """
    # Enviamos la peticion a la API
    url = u.url_files + "delete"
    args = {'file_id': file_id}
    print("Solicitando borrado del fichero #{}...".format(file_id), end="")
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("\nError de conexion")
        return

    if r.status_code == requests.codes.ok:
        print("OK")
        answer = r.json()
        file_id = answer.get('file_id')
        print("Fichero con ID#{} borrado correctamente".format(file_id))
    else:
        print()
        u.error(r)


def list_files():
    """
        Nombre: list_files
        Descripcion: Lista todos los ficheros pertenecientes al usuario.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos: Ninguno
        Retorno: Ninguno
    """
    # Enviamos la peticion a la API
    url = u.url_files + "list"
    print("Buscando ficheros del usuario en el servidor...", end="")
    try:
        r = requests.post(url, headers=u.headers)
    except requests.ConnectionError:
        print("\nError de conexion")
        return

    # Si es correcta impimimos todos los ficheros del usuario
    if r.status_code == requests.codes.ok:
        print("OK")
        answers = r.json()
        files_list = answers.get('files_list')
        num_files = answers.get('num_files')
        print("{} ficheros encontrados:".format(num_files))
        i = 0
        for file in files_list:
            i += 1
            print("[{}] ID#{} {}".format(i, file.get('fileID'), file.get('fileName')))
    else:
        print()
        u.error(r)


def prueba_files():
    # upload('383336', 'Prueba.txt')
    # delete('De29fbC4')
    list_files()
    download('6edD0F29', '383336')


users.py
import requests
from Cryptodome.PublicKey import RSA
import utils as u


def create_id(nombre, email):
    """
        Nombre: create_id
        Descripcion: Funcion que manda una peticion a la API para crear un usuario. Crea el par de claves RSA,
            guarda la privada y envía la publica a la API.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -nombre: nombre del usuario a crear.
            -email: email del usuario.
        Retorno: Ninguno
    """
    print("Creando un id en el sistema...")
    # Creamos el par de claves
    print("-> Generando par de claves RSA de 2048 bits...", end="")
    key = RSA.generate(2048)
    public_key = key.publickey().export_key()
    print("OK")

    # Enviamos la peticion a la API
    url = u.url_users + 'register'
    args = {'nombre': nombre, 'email': email, 'publicKey': public_key}
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("Error de conexion")
        return

    # Si la peticion es correcta guardamos la clave privada del usuario e imprimimos su ID
    if r.status_code == requests.codes.ok:
        f = open("clave.pem", "wb")
        f.write(key.export_key('PEM'))
        f.close()
        answer = r.json()
        user_id = answer.get('userID')
        print("Identidad con ID#{} creada correctamente".format(user_id))
    else:
        print()
        u.error(r)


def search_id(data_search):
    """
        Nombre: search_id
        Descripcion: Busca un usuario cuyo nombre o correo electronico contenga cadena en el repositorio
            de identidades de SecureBox, e imprime su ID.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -data_search: cadena a buscar.
        Retorno: Ninguno
    """

    # Enviamos la peticion a la API
    url = u.url_users + 'search'
    args = {'data_search': data_search}
    print("Buscando usuario '" + data_search + "' en el servidor...", end="")
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("\nError de conexion")
        return
    # Si es correcta imprimimos todas las coincidencias devueltas
    if r.status_code == requests.codes.ok:
        print("OK")
        answers = r.json()
        print("{} usuarios encontrados:".format(len(answers)))
        i = 0
        for answer in answers:
            print("[{}] {}, {}, ID: {}".format(i + 1, answer['nombre'], answer['email'], answer['userID']))
            i += 1
    else:
        print()
        u.error(r)


def delete_id(user_id):
    """
        Nombre: delete_id
        Descripcion: Borra la identidad con ID id registrada en el sistema. Solo puede borrar la suya propia.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -userID: ID a borrar.
        Retorno: Ninguno
    """

    # Envia la peticion a la API
    url = u.url_users + 'delete'
    args = {'userID': user_id}
    print("Solicitando borrado de la identidad #{}...".format(user_id), end="")
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("\nError de conexion")
        return

    # Si la peticion es correcta imprimimos el ID eliminado
    if r.status_code == requests.codes.ok:
        print("OK")
        answer = r.json()
        user_id = answer.get('userID')
        print("Identidad con ID#{} borrada correctamente".format(user_id))
    else:
        print()
        u.error(r)


def get_public_key(user_id):
    """
        Nombre: get_public_key
        Descripcion: Obtiene la clave pública de un usuario.
            En caso de error lo imprime por pantalla y sale de la funcion.
        Argumentos:
            -userID: ID a borrar.
        Retorno: la clave publica o None en caso de error.
    """

    # Enviamos la peticion a la API
    url = u.url_users + 'getPublicKey'
    args = {'userID': user_id}
    print("-> Solicitando clave de la identidad #{}...".format(user_id), end="")
    try:
        r = requests.post(url, headers=u.headers, json=args)
    except requests.ConnectionError:
        print("\nError de conexion")
        return None

    if r.status_code == requests.codes.ok:
        print("OK")
        return RSA.import_key(r.json().get('publicKey'))
    else:
        print()
        u.error(r)
        return None


def prueba():
    create_id("Carmen", "carmen.diezmenendez@estudiante.uam.es")
    search_id("Carmen")
    delete_id(38333336)


utils.py
import os
import configparser
from Cryptodome.Cipher import AES
from Cryptodome.Util.Padding import pad, unpad

# Variables globales
token = ""
path_archivos = ""
url_users = ""
url_files = ""
headers = {}


def config_ini(password, iv):
    """
        Nombre: config_ini
        Descripcion: Realiza la configuración inicial necesaria.
            Lee de ini.conf el token del usuario y el path para guardar archivos.
        Argumentos: Ninguno
        Retorno: Ninguno
    """
    global token, path_archivos, url_users, url_files

    f = open('ini.conf', 'rb')
    encriptado = f.read()
    cipher = AES.new(pad(password.encode('utf-8'), 32), AES.MODE_CBC, iv)
    contenido = unpad(cipher.decrypt(encriptado), AES.block_size)
    f.close()

    # Leemos token.txt y lo guardamos en la variable global token
    config = configparser.ConfigParser()
    config.read_string(contenido.decode('utf-8'))
    if 'token' in config['DEFAULT']:
        token = config['DEFAULT']['token']
    else:
        return

    # Guardamos el path de archivos y si no ponemos uno por defecto
    if 'path_archivos' in config['DEFAULT']:
        path_archivos = config['DEFAULT']['path_archivos']
    else:
        path_archivos = "Archivos/"

    # Guardamos los endpoints de la API
    if 'url_users' in config['DEFAULT']:
        url_users = config['DEFAULT']['url_users']
    else:
        return

    if 'url_files' in config['DEFAULT']:
        url_files = config['DEFAULT']['url_files']
    else:
        return

    # Si no existe path_archivos creamos la carpeta
    basedir = os.path.dirname(path_archivos)
    if not os.path.exists(basedir):
        os.makedirs(basedir)


def error(request):
    """
        Nombre: error
        Descripcion: Imprime el mensaje de error correspodiente recibido por una
            request.
        Argumentos:
            - request: respuesta de la peticion al servidor.
        Retorno: Ninguno
    """
    json = request.json()

    # Si el error esta definido imprimimos la informacion completa
    if "error_code" in json:
        print("Error {}: {} -> {}".format(json.get("http_error_code"), json.get("error_code"), json.get("description")))
    else:
        print("Error {}".format(request.status_code))


crypto.py
from Cryptodome.Signature import pkcs1_15
from Cryptodome.Hash import SHA256
from Cryptodome.PublicKey import RSA
from Cryptodome.Random import get_random_bytes
from Cryptodome.Cipher import AES
from Cryptodome.Util.Padding import pad, unpad
from Cryptodome.Cipher import PKCS1_OAEP


def encrypt(mensaje, clave_pub_r):
    """
        Nombre: encrypt
        Descripcion: Funcion que encripta un mensaje mediante la AES con modo de encadenamiento CBC.
            Para ello crea una clave simetrica y un vector de inicializacion aleatorios. Genera
            el sobre digital de la clave.
        Argumentos:
            -mensaje: mensaje a cifrar.
            -clave_pub_r: clave publica del receptor.
        Retorno: IV + sobre digital + mensaje cifrado
    """
    print("-> Encriptando el mensaje...", end="")
    # Cifrado simétrico: AES con modo de encadenamiento CBC,
    # con IV de 16 bytes, y longitud de clave de 256 bits.
    # Generamos una clave de 256 bits = 32 bytes
    iv = get_random_bytes(16)
    clave_s = get_random_bytes(32)
    cipher = AES.new(clave_s, AES.MODE_CBC, iv)
    # El mensaje tiene que ser multiplo del tamanio del bloque
    # (16 en AES) asi que añadimos padding
    mensaje_cifrado = cipher.encrypt(pad(mensaje, AES.block_size))

    # Obtenemos el sobre con OAEP
    # La clave es clave_pub_r y el mensaje es clave_s
    cipher = PKCS1_OAEP.new(clave_pub_r)
    try:
        sobre_digital = cipher.encrypt(clave_s)
        print("OK")
        return iv + sobre_digital + mensaje_cifrado
    except ValueError:
        print("Error al cifrar.")
        return None


def sign(mensaje):
    """
        Nombre: sign
        Descripcion: Funcion que obtiene la firma de un mensaje mediante
            PCKS usando la clave privada del emisor (clave.pem).
        Argumentos:
            -mensaje: mensaje a firmar.
        Retorno: firma
    """
    print("-> Generando la firma...", end="")
    # Hacemos hash256 del mensaje
    h = SHA256.new(mensaje)
    # Ciframos con PKCS 1.5 el hash con la clave privada del emisor
    try:
        # Obtenemos la clave privada
        f = open('clave.pem', 'r')
        clave_priv_e = RSA.import_key(f.read())
        f.close()

        # sign puede generar ValueError y TypeError
        firma = pkcs1_15.new(clave_priv_e).sign(h)
        print("OK")
        return firma
    except (ValueError, TypeError):
        print("\nError al firmar.")
        return None


def enc_sign(mensaje, clave_pub_r):
    """
        Nombre: enc_sign
        Descripcion: Funcion que firma y cifra un mensaje en bytes.
        Argumentos:
            -mensaje: mensaje a firmar y cifrar
            -clave_pub_r: clave publica del receptor
        Retorno: mensaje firmado y cifrado
    """
    # Obtenemos la firma
    firma_digital = sign(mensaje)
    if firma_digital is None:
        return None

    # Ciframos la firma digital con el mensaje
    return encrypt(firma_digital + mensaje, clave_pub_r)


def decrypt(mensaje):
    """
        Nombre: decrypt
        Descripcion: Funcion que descifra un mensaje.
        Argumentos:
            -mensaje: mensaje cifrado y firmado
        Retorno: mensaje descifrado
    """
    # Obtenemos el vector de inicializacion y el sobre digital
    print("-> Descifrando fichero...", end="")
    iv = mensaje[0:16]
    sobre_digital = mensaje[16:16 + 256]

    # Obtenemos la clave privada del receptor
    f = open('clave.pem', 'r')
    clave_priv_r = RSA.import_key(f.read())
    f.close()

    # Obtenemos el sobre con OAEP
    cipher = PKCS1_OAEP.new(clave_priv_r)
    try:
        clave_s = cipher.decrypt(sobre_digital)
    except ValueError:
        print("Error: No es posible descifrar")
        return None

    # Desciframos y quitamos el pad
    cipher = AES.new(clave_s, AES.MODE_CBC, iv)
    mensaje_descifrado = unpad(cipher.decrypt(mensaje[16 + 256:]), AES.block_size)

    print("OK")
    return mensaje_descifrado


def check_sign(mensaje_descifrado, clave_pub_e):
    """
        Nombre: check_sign
        Descripcion: Funcion que comprueba la firma y obtiene el mensaje original.
        Argumentos:
            -mensaje: mensaje firmado
        Retorno: mensaje original
    """
    print("-> Verificando firma...", end="")

    firma_digital = mensaje_descifrado[:256]
    mensaje_original = mensaje_descifrado[256:]

    # Hacemos hash256 del mensaje
    h = SHA256.new(mensaje_original)

    # Comprobamos la firma_digital con el hash
    try:
        pkcs1_15.new(clave_pub_e).verify(h, firma_digital)
    except ValueError:
        print("Error al verificar la firma digital.")
        return None
    print("OK")
    return mensaje_original


def prueba_crypto():
    new_key = RSA.generate(2048)
    f = open('clave.pem', 'wb')
    f.write(new_key.export_key())
    f.close()
    public_key = new_key.publickey()
    encriptado = enc_sign("Hola mundo".encode('utf-8'), public_key)
    mensaje_descifrado = decrypt(encriptado)
    print(check_sign(mensaje_descifrado, public_key).decode("utf-8"))


securebox_client.py
import argparse
import files
import users
import crypto
import utils as u
import os
import sys
import getpass
from Cryptodome.Cipher import AES
from Cryptodome.Random import get_random_bytes
from Cryptodome.Hash import SHA256
from Cryptodome.Util.Padding import pad

if __name__ == "__main__":
    # Pedimos una contraseña si es la primera vez que se ejecuta securebox,
    # es decir, no existe fichero de hash con la contraseña
    password = getpass.getpass(prompt="Introduce tu contraseña: ")
    if not os.path.exists("hash.txt"):
        f = open("hash.txt", "wb")
        b_password = pad(password.encode('utf-8'), 32)
        iv = get_random_bytes(16)

        cipher = AES.new(b_password, AES.MODE_CBC, iv)
        f.write(iv + SHA256.new(b_password).digest())
        f.close()

        f = open("ini.conf", "rb+")
        config = f.read()
        f.truncate(0)
        f.close()

        encriptado = cipher.encrypt(pad(config, AES.block_size))

        f = open("ini.conf", "wb")
        f.write(encriptado)
        f.close()
    else:
        f = open("hash.txt", "rb")
        contenido = f.read()
        iv = contenido[0:16]
        h = contenido[16:]
        hash_password = SHA256.new(pad(password.encode('utf-8'), 32)).digest()

        if h != hash_password:
            print("Contraseña incorrecta.")
            sys.exit(0)

    # Guardamos el token, el header para todas las peticiones y el path de archivos
    # Estas variables están en utils
    u.config_ini(password, iv)
    u.headers['Authorization'] = "Bearer " + u.token

    # Si el token existe continuamos
    if u.token != "":

        # Añadimos todos los posibles argumentos al parser
        parser = argparse.ArgumentParser(description='Interactúa con securebox')

        parser.add_argument('--create_id',
                            help='Crea una nueva identidad (par de claves púlica y privada) para un usuario con nombre '
                                 'nombre y correo email, y la registra en SecureBox, para que pueda ser encontrada por '
                                 'otros usuarios',
                            nargs=2, metavar=('Nombre', 'Email'))

        parser.add_argument('--search_id',
                            help='Busca un usuario cuyo nombre o correo electrónico contenga cadena en el repositorio '
                                 'de identidades de SecureBox, y devuelve su ID',
                            metavar='Nombre/Email')

        parser.add_argument('--delete_id', help='Borra la identidad con ID id registrada en el sistema', metavar='id')

        parser.add_argument('--upload',
                            help='Envia un fichero a otro usuario, cuyo ID es especificado con la opción --dest_id',
                            metavar='fichero')

        parser.add_argument('--source_id', help='ID del emisor del fichero', metavar='id_emisor')

        parser.add_argument('--dest_id', help='ID del receptor del fichero', metavar='id_receptor')

        parser.add_argument('--list_files', action='store_true', help='Lista todos los ficheros pertenecientes al '
                                                                      'usuario')

        parser.add_argument('--download', help='Recupera un fichero con ID id_fichero del sistema',
                            metavar='id_fichero')

        parser.add_argument('--delete_file', help='Borra un fichero del sistema', metavar='id_fichero')

        parser.add_argument('--encrypt',
                            help='Cifra un fichero, de forma que puede ser descifrado por otro usuario, cuyo ID es '
                                 'especificado con la opción --dest_id',
                            metavar='fichero')

        parser.add_argument('--sign', help='Firma un fichero', metavar='fichero')

        parser.add_argument('--enc_sign', help='Cifra y firma un fichero para --dest_id', metavar='fichero')

        parser.add_argument('--dec_check', help='Desencripta y valida la firma de un fichero de --source_id',
                            metavar='fichero')

        args = parser.parse_args()

        # Se realiza la accion correspondiente segun los argumentos
        if args.create_id:
            users.create_id(args.create_id[0], args.create_id[1])

        elif args.search_id:
            users.search_id(args.search_id)

        elif args.delete_id:
            users.delete_id(args.delete_id)

        elif args.upload:
            if args.dest_id:
                files.upload(args.upload, args.dest_id)
            else:
                print("--upload necesita --dest_id")

        elif args.list_files:
            files.list_files()

        elif args.download:
            if args.source_id:
                files.download(args.download, args.source_id)
            else:
                print("--download necesita --source_id")

        elif args.delete_file:
            files.delete(args.delete_file)

        # Para encriptar pedimos la clave publica del destinatario, ciframos
        # y guardamos el fichero encriptado en path_archivos
        elif args.encrypt:
            if args.dest_id:
                publicKey = users.get_public_key(args.dest_id)
                if publicKey is not None:
                    try:
                        f = open(args.encrypt, "rb")
                        mensaje_cifrado = crypto.encrypt(f.read(), publicKey)
                        f.close()
                        # Si se ha cifrado correctamente lo guardamos como: enc_ + nombre del fichero
                        if mensaje_cifrado is not None:
                            file_name = os.path.basename(args.encrypt)
                            f = open(u.path_archivos + "enc_" + file_name, "wb")
                            f.write(mensaje_cifrado)
                            f.close()
                    except FileNotFoundError:
                        print("El archivo no existe")
            else:
                print("--encrypt necesita --dest_id")

        # Firmamos un fichero y la guardamos en path_archivos
        elif args.sign:
            try:
                f = open(args.sign, 'rb')
                mensaje = f.read()
                firma_digital = crypto.sign(mensaje)
                f.close()
                # Si se ha firmado correctamente guardamos la firma y el mensaje como: sign_ + nombre del fichero
                if firma_digital is not None:
                    file_name = os.path.basename(args.sign)
                    f = open(u.path_archivos + "sign_" + file_name, "wb")
                    f.write(firma_digital + mensaje)
                    f.close()

            except FileNotFoundError:
                print("El archivo no existe")

        # Encripta y frima un archivo para un destinatario y lo guarda en path_archivos
        elif args.enc_sign:
            if args.dest_id:
                try:
                    f = open(args.enc_sign, 'rb')
                    mensaje = f.read()
                    f.close()
                    publicKey = users.get_public_key(args.dest_id)
                    if publicKey is not None:
                        mensaje_enc_sign = crypto.enc_sign(mensaje, publicKey)
                        # Si se ha realizado correctamente lo guardarmos como: enc_sign_ + nombre del fichero
                        if mensaje_enc_sign is not None:
                            file_name = os.path.basename(args.enc_sign)
                            f = open(u.path_archivos + "enc_sign_" + file_name, "wb")
                            f.write(mensaje_enc_sign)
                            f.close()
                except FileNotFoundError:
                    print("El archivo no existe")
            else:
                print("--enc_sign necesita --dest_id")

        # Desencripta y comprueba la firma de un fichero dado. Guarda el fichero
        # dado en path_archivos
        elif args.dec_check:
            if args.source_id:
                try:
                    f = open(args.dec_check, 'rb')
                    mensaje = f.read()
                    f.close()
                    mensaje_descifrado = crypto.decrypt(mensaje)
                    publicKey = users.get_public_key(args.source_id)
                    if publicKey is not None:
                        check = crypto.check_sign(mensaje_descifrado, publicKey)
                        print("La firma es ", end="")
                        if check is None:
                            print("incorrecta")
                        else:
                            print("correcta")
                            file_name = os.path.basename(args.dec_check)
                            f = open(u.path_archivos + "dec_check_" + file_name, "wb")
                            f.write(check)
                            f.close()
                except FileNotFoundError:
                    print("El archivo no existe")
            else:
                print("--dec_check necesita --source_id")

        else:
            print("Introduce --help para ver los posibles comandos")
    else:
        print("No hay token.")
